graylog2-server/src/test/java/org/graylog2/plugin/MessageTest.java

```diff
@@ -669,4 +673,65 @@ public void testMetadataDefault() throws NoSuchFieldException, IllegalAccessExce
         assertThat(message.getMetadataValue("badKey", "default")).isEqualTo("default");
         assertThat(message.getMetadataValue("stateKey", "default")).isEqualTo(10L);
     }
+
+    @Test
+    public void testTimestampConversionWithWrongDate() {
+        // Do not use fixed time from setUp() in this test
+        DateTimeUtils.setCurrentMillisSystem();
+
+        final Message message = new Message("message", "source", Tools.nowUTC().minusMinutes(2));
+        final DateTime previousTimestamp = message.getTimestamp();
+
+        message.addField(Message.FIELD_TIMESTAMP, 1234);
+
+        assertThat(message.getTimestamp()).isInstanceOf(DateTime.class);
+        // got replaced by a current timestamp
+        assertThat(message.getTimestamp()).isNotEqualTo(previousTimestamp);
+
+        assertThat(message.processingErrors()).satisfies(e -> {
+            assertThat(e).hasSize(1);
+            assertThat(e.get(0).getCause()).isEqualTo(ProcessingFailureCause.InvalidTimestampException);
+            assertThat(e.get(0).getMessage()).startsWith("Invalid type for field timestamp 'Integer'");
+        });
+    }
+
+    @Test
+    public void testTimestampConversionWithNullDate() {
+        // Do not use fixed time from setUp() in this test
+        DateTimeUtils.setCurrentMillisSystem();
+
+        final Message message = new Message("message", "source", Tools.nowUTC().minusMinutes(2));
+        final DateTime previousTimestamp = message.getTimestamp();
+
+        message.addField(Message.FIELD_TIMESTAMP, null);
+
+        assertThat(message.getTimestamp()).isInstanceOf(DateTime.class);
+        // got replaced by a current timestamp
+        assertThat(message.getTimestamp()).isNotEqualTo(previousTimestamp);
+
+        assertThat(message.processingErrors()).satisfies(e -> {
+            assertThat(e).hasSize(1);
+            assertThat(e.get(0).getCause()).isEqualTo(ProcessingFailureCause.InvalidTimestampException);
+            assertThat(e.get(0).getMessage()).startsWith("null value for field timestamp");
+        });
+    }
+
+    @Test
+    public void testTimestampConversionWithLocalDateTime() {
+        // Do not use fixed time from setUp() in this test
+        DateTimeUtils.setCurrentMillisSystem();
+
+        final Message message = new Message("message", "source", Tools.nowUTC().minusMinutes(2));
+        final LocalDateTime localDate = LocalDateTime.of(2021, Month.AUGUST, 19, 12, 0);
+        final ZonedDateTime zonedDateTime = ZonedDateTime.of(localDate, ZoneOffset.UTC);
+
+        message.addField(Message.FIELD_TIMESTAMP, zonedDateTime);
+        assertThat(message.getTimestamp()).isInstanceOf(DateTime.class);
+        final DateTimeFormatter formatter = DateTimeFormat.forPattern("yyyy-MM-dd HH:mm").withZone(DateTimeZone.UTC);
+        final DateTime expectedLocalDateEquivalent = formatter.parseDateTime("2021-08-19 12:00");
```

Comment:
Why the intermediate step with string parsing is need here?

Reply:
Seemed the easiest way to construct an equivalent DateTime to me

Reply:
changed that